### DISCUSSION TITLE
Partial HTML pre-render in compile-time

### DISCUSSION THREAD

**Discussion by cawa-93** [vuejs/rfcs#396] [General]
I was wondering why and is it possible at all to partially render some of the components in project "compile-time"?

I'd better explain with an example. A typical template looks something like this:

```html
<!-- index.html -->
<body>
  <div id="app"></div>
</body>
```
```vue
<!-- App.vue -->
<app-header/>
  
<main>
  <router-view/>
</main>

<app-footer/>
```

Note that I'm not talking about server side rendering, but generation during build with tools like webpack or vite. 

So now these tools generate an empty root element and JS which creates content after loading. 

```html
<body>
  <div id="app"></div>
  <script module src="main.js"></script>
</body>
```

However, there are scenarios where some part of the pattern from the root element to the depth is mostly static. Why not render HTML in advance, and in runtime just hydrate these parts?

In my example, the generated HTML might look something like this:

```html
<body>
  <div id="app">
    <header> ... header content </header>
  
    <main></main>

    <footer> ... footer content </footer>
  </div>
  <script module src="main.js"></script>
</body>
```

This could be a very big improvement on the first rendering.

**Comment by posva** [comment]
At that point, you could do SSG: https://github.com/antfu/vite-ssg which will give you the perf boost you want while still having a static server

**Comment by ElMassimo** [comment]
@cawa-93 In certain use cases you can try using [îles](https://github.com/ElMassimo/iles), which will prerender all components and hydrate only the ones you specify should be interactive.

---

Supporting partial hydration in Vue core would be very powerful and unlock the same benefits for a wider range of use cases.

**Comment by patchthecode** [comment]
so.. can Vue3 do this or not?
and if not vue, then vite?
Nuxt might be able to do it, but my app is a very large vue app and I wont be able to switch like this.
